### ISSUE TITLE
Add support for launching AMD SEV-SNP enabled guest

### ISSUE THREAD

**Issue by jinankjain** [cloud-hypervisor/cloud-hypervisor#4761]
This is a new feature request as part of this request, we would like to extend CloudHypervisor with the capability to launch an AMD SEV-SNP enlightened guest.

Whitepaper for SEV-SNP feature: https://www.amd.com/system/files/TechDocs/SEV-SNP-strengthening-vm-isolation-with-integrity-protection-and-more.pdf

We should create a platform feature called `snp` which would indicate if guests were SEV-SNP enabled or not. This is something along the line of platform feature `tdx`.

Along with it I would also propose to have a feature flag to guard all SEV-SNP related changes until all the API changes are finalized. This is again in line with tdx feature.

List of task: 

- [x] SNP feature guard
- [x]  IGVM parser and IGVM feature guard
- [x]  Support IGVM file as a payload
- [x]  Make memory/arch module to support IGVM file
- [x]  Change IO path to support Sev-SNP
- [x]  Handle GHCB Exits in hypervisor crates

CC: @rbradford @liuw

**Comment by rbradford** [comment]
Are you implementing this only on KVM?

**Comment by jinankjain** [comment]
Current focus is on MSHV. In future we can think about KVM.

**Comment by jinankjain** [comment]
@rbradford @liuw Currently we have an issue with cloud-hypervisor vcpu's creation on MSHV when we try to launch a SEV-SNP enabled VM. In the current approach, we allocate and map the guest's memory first and then allocate vCPUs. But MSHV requires vCPUs creation before allocating and mapping guest memory. This implies CpuManager should be initialized before MemoryManager.

I have v1 proposed for the aforementioned problem in #4957 but I am open to other alternatives as well. I would love to hear the community's opinion on that.

**Comment by rbradford** [comment]
So I looked at *why* CpuManager needed the MemoryManager:

1.) The guest_memory for programming the GDT when booting a vCPU (not needed for resizing). Can be refactored to be past in when the vCPUs are configured from the `Vm::boot` function
2.) For `guest_debug` feature for reading/writing guest memory. Again this can be passed in from the code in the `vm` module.
3.) For populating the the CPUID when running on SGX. This can be refactored to generate the common cpuid later.

The first 2 are nice general refactoring that reduce the code coupling and overall make the project cleaner. The last one is a little contrived for solving this specific issue but not as much complexity compared to your PR in #4957.

You can see my WIP on this at https://github.com/rbradford/cloud-hypervisor/tree/2022-12-01-decouple-cpu-manager-memory-manager which handles everything but the delayed CPUID initialisation.

I propose to add a new method `CpuManager::generate_common_cpuid()` which would take a `&MemoryManager` and generate the common CPUID state after the CpuManager and MemoryManager were created but before any vCPUs were configured (which is where the CPUID state is needed.)

**Comment by jinankjain** [comment]
@rbradford This sounds like a very good plan and indeed a much cleaner solution as compared to mine. Are you planning to continue working on this or should I take it forward?

**Comment by rbradford** [comment]
> @rbradford This sounds like a very good plan and indeed a much cleaner solution as compared to mine. Are you planning to continue working on this or should I take it forward?

Yes, i'm just waiting for #4963 to land.

**Comment by rbradford** [comment]
You need to outline how this feature will work and how it will be structured. As we discussed earlier it will be need to be under a compile time feature flag as we have no testing for it.

**Comment by jinankjain** [comment]
As far the feature is concerned it would be under a feature flag "snp". But currently we are working on the general refactoring which is not specific to SEV-SNP feature itself. The current problem that we have at hand is related to vCPU creation. In the current, workflow vCPUs are created after we have mapped the guest memory. But according to SEV-SNP specification vCPUs must be created before mapping any memory to guest because usually the guest memory would be encrypted. This translates to creating CpuManager independent of MemoryManger. Thus we would be able call `create_boot_cpus` before initializing the MemoryManager. @rbradford Does this make requirements a little bit more clear?

**Comment by rbradford** [comment]
Seems reasonable to split that functionality.

**Comment by jinankjain** [comment]
> > But according to SEV-SNP specification vCPUs must be created before mapping any memory to guest because usually the guest memory would be encrypted. This translates to creating CpuManager independent of MemoryManger. Thus we would be able call `create_boot_cpus` before initializing the MemoryManager. @rbradford Does this make requirements a little bit more clear?
> 
> This is not sufficient detail. You need to elaborate more. Do you mean that the vCPU FDs must opened or do you mean that they must be running? Do you have a pointer to the MSHV kernel API in question? Perhaps you could prototype a simple VMM using the basic vm-memory, mshv-ioctls and linux-loader crates?

Coming back to this discussion I think I have a bit more clarity on this front too. To answer your question first: yes, I want the vCPU FDs to open first but not running. Or if we are doing the refactoring as I mentioned above, then we can get away with this requirement of untangling CpuManger from MemoryManger. Because then we would be creating the userspace mappings before starting the vCPUs that means we would have already opened the vCPU fd as part of `configure_vcpu`  call. In the end I wanted userspace mappings to be created after we have opened the vCPU fds.

**Comment by jinankjain** [comment]
> Seems reasonable to split that functionality.

Awesome then let me get to creating the pull request for this one.

**Comment by waywardgeek** [comment]
Hi, folks!  I am very interested in helping out with this effort.  Let me know how to get involved.  Thanks!

**Comment by liuw** [comment]
> Hi, folks! I am very interested in helping out with this effort. Let me know how to get involved. Thanks!

Hi Bill. I assume you're interested in KVM mostly? The work we (Microsoft) do here mostly cater to our own hypervisor. That said, it is in our best interest to make the code work on KVM, too. I do expect there will be deviation between KVM and MSHV though since the APIs exposed from the hypervisors are different.

If you can start with describing what your end goals are, we can perhaps talk more concretely. In the meantime, your review and input will also be very welcomed when we post our code.

**Comment by waywardgeek** [comment]
Hi, Liu.  Yes, we need SEV-SNP support for CHV running under KVM.

I am tasked with getting CHV to support reverse-sandboxing modes, and SEV-SNP in particular, for Google internal security purposes.  I specialize in building out new functionality rapidly.  Ideally, we need this working in under 12 months, and preferably closer to six.  For me it is a new language (Rust), and a new area of technology (VMMs), but that's pretty normal for me.  I'm looking forward to the work.

**Comment by liuw** [comment]
> Hi, Liu. Yes, we need SEV-SNP support for CHV running under KVM.
> 
> I am tasked with getting CHV to support reverse-sandboxing modes, and SEV-SNP in particular, for Google internal security purposes. I specialize in building out new functionality rapidly. Ideally, we need this working in under 12 months, and preferably closer to six. For me it is a new language (Rust), and a new area of technology (VMMs), but that's pretty normal for me. I'm looking forward to the work.

Very cool. I look forward to seeing your work.

**Comment by waywardgeek** [comment]
If you have tips for me, e.g. link to style guidelines, onboarding docs, let me know.

**Comment by liuw** [comment]
> If you have tips for me, e.g. link to style guidelines, onboarding docs, let me know.

Sorry for the delay in reply.

For styling we just use `rustfmt`. You can configure your editor to invoke it on every save.

The documents are under `docs`.

**Comment by waywardgeek** [comment]
Great, thanks.  I'm just starting to work on this now today.  I don't have to start from scratch.  [Enarx](https://github.com/enarx/enarx) has > 60 CLs related to enabling SEV-SNP, and they read line a detailed plan for enabling SEV-SNP support.  There's also I think 53 PRs in CHV that mention TDX.  I think the TDX support PRs (mostly) show where things need to be changed (though we need KVM).  Between the two sets of PRs I think I can figure this out.

I apologize for being a Rust, KVM, and CHV noob.  I'm going to try and get some readability reviews for Rust PRs internally before posting PRs here, or if you like I can post my worst versions here and have other folks I work with do the initial Rust readability reviews here in the open.

I appreciate any feedback, and have very thick skin, so just tell it like it is in reviews.

I haven't written any design doc, not even a 1-pager.  Maybe that would be a good place to start engaging with the CHV team?  The solution space seems large, and I'd like to choose the right approach.  For example, the Enarx code created a "Backend" abstraction that can be used for SEV or SGX.  One approach would be to find the commonality between TDX support and SEV/SEV-ES/SEV-SNP, and create traits for these confidential backends.  Alternatively, I could follow the style of the TDX code without creating a confidential backend abstraction.

I literally finished reading the Rust Book yesterday, and it will take me a few days to become familiar with the CHV and Enarx code bases.  Let me know if you'd like a 1-pager, something a bit more fleshed out, or if you have any insights for how this code should be structured.  Thanks in advance!

**Comment by Lencerf** [comment]
Hi folks, any updates about the SEV support in cluod-hypervisor?  I also see that AMD is in the list of supporters (https://www.cloudhypervisor.org/). Do you know if AMD has any plans of implementing this feature for cloud-hypervisor? Thanks!

**Comment by waywardgeek** [comment]
Unfortunately my effort was redirected almost as soon as it started.
That's just life working at any big company like the ones most of us work
for.  If it makes any difference I think that leadership got it wrong this
time.  The best technical solution was to enlighten cloud hypervisor.  I
suspect we'll be back when it becomes clear why this is the correct
technical solution, but of course I do not speak for Google.

**Comment by jinankjain** [comment]
I have started some initial work from MSHV side in PR: #5628

**Comment by russell-islam** [comment]
https://github.com/cloud-hypervisor/cloud-hypervisor/pull/5846 covers task 2 and 3

**Comment by russell-islam** [comment]
Another PR for task 2/3 https://github.com/cloud-hypervisor/cloud-hypervisor/pull/5872

**Comment by liuw** [comment]
@jinankjain @russell-islam I think this issue can be closed now?

**Comment by waywardgeek** [comment]
Maybe leave it open for a while?  There may be PRs in 2025 adding similar support for KVM.

**Comment by liuw** [comment]
@waywardgeek there is a separate issue for that https://github.com/cloud-hypervisor/cloud-hypervisor/issues/6653

**Comment by jinankjain** [comment]
@liuw Indeed this can be closed since the MSHV support is merged to mainline.

**Comment by liuw** [comment]
Closing this. KVM support is tracked separately.
